### PR TITLE
[feat-2134][framework]: Rework the error management code in the Backend

### LINKED ISSUE #2134: Add more context to error messages
## What and why to refactor
Many plugin subtasks directly return errors without adding any extra context to the message. This makes troubleshooting difficult. For example, [this](https://github.com/apache/incubator-devlake/blob/0d3aa41e5e3db60e7ddfd3c3835f40e8c1c8eeb4/plugins/github/tasks/pr_commit_extractor.go#L85), which might produce a Json unmarsal error, needs to have a message that additionally tells us something like "unable to get Input while extracting pull request commits".

## Describe the solution you'd like
Identify all such plugin subtasks and wrap the error messages with more context.

## Related issues
Similar to #1447 
## Additional context
n/a

### PR DESCRIPTION
**Pull request by keon94**
### ⚠️ &nbsp;&nbsp;Pre Checklist

> Please complete _ALL_ items in this checklist, and remove before submitting

- [x] I have read through the [Contributing](https://devlake.apache.org/community/) Documentation & [PR Template](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] This PR is using a `label` (bug, feature etc.)
- [x] My code is has necessary documentation (if appropriate)
- [x] I have added any relevant tests
- [ ] This section (**⚠️ &nbsp;&nbsp;Pre Checklist**) will be removed when submitting PR

# Summary

### Does this close any open issues?
Closes #2134

### Screenshots
Include any relevant screenshots here.

### Other Information
Blocks #2907

### CODE REVIEW

**Comment by keon94** [comment]
@hezyin @klesh @Startrekzky 

I've been working on improving our error management infrastructure, and have implemetend a work-in-progress implementation of my ideas. Obviously, I'll have to go and find all the places where we are throwing errors and pass them through this new API. But I'd like to get your input as well before I do too much, so please have a look.

Here's a screenshot of a temporary test suite I wrote:
image

And the results (the errors printed)
image

File: errors/impl.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+package errors
+
+import "fmt"
+import cerror "github.com/cockroachdb/errors"
+
+type (
+	crdbErrorImpl struct {
+		wrappedRaw error
+		wrapped    *crdbErrorImpl
```
Comment (keon94) [inline errors/impl.go:26]: we can potentially split this, into two strings: msg and usrMsg, and have the consumer code specify a custom User msg too optionally. The UserMessage() API would only then return the usrMsg fields if they are not empty

File: errors/impl.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+func (e *crdbErrorImpl) Error() string {
+	return fmt.Sprintf("%+v", e.wrappedRaw)
+}
+
+func (e *crdbErrorImpl) Message() string {
+	return e.wrappedRaw.Error()
+}
+
+func (e *crdbErrorImpl) UserMessage() string {
```
Comment (mindlesscloud) [inline errors/impl.go:40]: It seems that the body of this function cloud be replaced by one line of code:
```go
return strings.Join(e.getUserMessages(), "\ncaused by: ")
```

**Comment by keon94** [comment]
Example after including the optional "user-friendly" feature
image
image

**Comment by keon94** [comment]
Example backend log message with "test-jira-connection" erroring out
image

Response as seen by the front-end
image

**Comment by keon94** [comment]
@e2corporation with my new error messaging scheme [here](https://user-images.githubusercontent.com/25063936/188023608-1b5c6167-9869-4602-8fd1-e3909d821b6c.png) I am returning a multi-line cause-hierarchy of messages on the response. Let me know your thoughts. cc @klesh

**Review by klesh — APPROVED** [review]
LGTM

**Pull request merged**